commit 6f6943fc94ada145c0af5f62f719abad73825e5b
Author: Tom Lane <tgl@sss.pgh.pa.us>
Date:   Tue Jan 11 14:22:00 2022 -0500

    Improve error message for missing extension.
    
    If we get ENOENT while trying to read an extension control file,
    report that as a missing extension (with a HINT to install it)
    rather than as a filesystem access problem.  The message wording
    was extensively bikeshedded in hopes of pointing people to the
    idea that they need to do a software installation before they
    can install the extension into the current database.
    
    Nathan Bossart, with review/wording suggestions from Daniel
    Gustafsson, Chapman Flack, and myself
    
    Discussion: https://postgr.es/m/3950D56A-4E47-48E7-BF9B-F5F22E268BE7@amazon.com

diff --git a/src/backend/commands/extension.c b/src/backend/commands/extension.c
index 345787fe2ce..a2e77c418a4 100644
--- a/src/backend/commands/extension.c
+++ b/src/backend/commands/extension.c
@@ -485,15 +485,26 @@ parse_extension_control_file(ExtensionControlFile *control,
 	else
 		filename = get_extension_control_filename(control->name);
 
 	if ((file = AllocateFile(filename, "r")) == NULL)
 	{
-		if (version && errno == ENOENT)
+		if (errno == ENOENT)
 		{
-			/* no auxiliary file for this version */
-			pfree(filename);
-			return;
+			/* no complaint for missing auxiliary file */
+			if (version)
+			{
+				pfree(filename);
+				return;
+			}
+
+			/* missing control file indicates extension is not installed */
+			ereport(ERROR,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+					 errmsg("extension \"%s\" is not available", control->name),
+					 errdetail("Could not open extension control file \"%s\": %m.",
+							   filename),
+					 errhint("The extension must first be installed on the system where PostgreSQL is running.")));
 		}
 		ereport(ERROR,
 				(errcode_for_file_access(),
 				 errmsg("could not open extension control file \"%s\": %m",
 						filename)));
